configuration_broker/subscriber1.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+// Contributed by Configured Things Ltd
+
+#include <compartment.h>
+#include <cstdlib>
+#include <debug.hh>
+#include <fail-simulator-on-error.h>
+#include <thread.h>
+
+// Define a sealed capability that gives this compartment
+// read access to configuration data "config1"
+#include "config_broker.h"
+#include "futex.h"
+
+// Config Item this subscriber is interested in
+#define CONFIG_ITEM_NAME "config1"
+DEFINE_READ_CONFIG_CAPABILITY(CONFIG_ITEM_NAME)
+
+// Expose debugging features unconditionally for this compartment.
+using Debug = ConditionalDebug<true, "Subscriber #1">;
+
+#include "data.h"
+#include "sandbox.h"
+
+// Current config value
+Data *configData = nullptr;
+
+//
+// Process a change in config data
+//
+void process_update(ConfigItem *config)
+{
+	if (config->data != nullptr)
+	{
+		if (sandbox_validate(config->data) < 0)
```

Comment:
Why - my thinking was that its the subscriber that owns any knowledge of the meaning of the data and how it should be validated.  Even the publisher only has to know what items exist and how big they are (they could be just blobs from mqtt topics).

Reply:
If validation is needed for a node type, requiring validation in every consumer means that it’s easy for one to forget. If validation is needed but isn’t done until the subscribe point, we allow invalid data to persist. I think the API I’d want is to register the validation functions for each node, so the broker will accept any data that meets the validation requirements but guarantees that it only ever shares validated data.

Reply:
I wanted the decision on what validation mans to rest with each subscriber; they are the things that ultimately have to establish trust the data and it it could be that two subscribed for the same item have different requirements on what that is, so there's not really a binary concept of valid or invalid.

Reply:
@davidchisnall  Would like to get your view on the following options around how/if to validate configuration values please

option 1 (pretty much what I have atm)
Data is defined as a struct. Publisher provides values into heap allocated by Broker during the callback.  Broker runs an independently registered validator against the data before deciding whether to accept / publish it or not. 

option 2: (nae use of Classes rather that struct)
Data is defined as a class, with setter / getter methods, with the validation defined in the setters. Flow is pretty much the same, but in the publisher callback we do a placement new and there is no need for a validation step as that's handled in the class methods.

Clearly option 2 depends on the publisher using the Class when it sets the values, but since we have an auditable view of which compartments have the capability to publish the item is it being overly paranoid to protect against some compromise in the publisher causing it to do something else in the callback ?  

I have a natural inclination to paranoia / showing how granular the trust model can be made with Cheri (hence option 1) but I'm wondering if in practice this is making things overly complex.  

I looked through the tests / examples & network-stack but couldn't spot any other examples of passing objects around to see what's preferred here in terms of using Classes vs just working with structs

Reply:
I'm not sure I understand 2.  You can't shouldn't pass objects with virtual functions across compartment boundaries (the vtable pointers will be in the caller's PCC and everything will break).

Reply:
Doh of course.  I was thinking that if I did a placement new into the heap space allocated by the Broker then I could access it as an object instance from any compartment that included the code that defined the object. 

I'll stick with using a library then - thanks.